### ISSUE TITLE
Something amiss with VimL parser (missing :endif in plugin)

### ISSUE THREAD

**Issue by aktau** [neovim/neovim#6431]
- `nvim --version`:  NVIM v0.2.0-1161-g6afa7d66
- Vim (version: ) behaves differently? Yes, it doesn't error

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 07 2017 03:33:23)
Included patches: 1-197, 322, 377-378

- Operating system/version: Linux
- Terminal name/version: urxvt
- `$TERM`: rxvt-unicode

### Actual behaviour

nvim prints the following before painting the editor window. This seemingly has to do with https://github.com/xolox/vim-misc which is required by https://github.com/xolox/vim-easytags. The error has not been printed by vanilla Vim up until now. Sadly, it's not 100% reproducible here (as I just noticed, after collecting a lot of info). I'm still keeping the text I gathered around, so ti can be referred to later.

```
Error detected while processing function xolox#easytags#initialize[33]..xolox#misc#path#which:
line   25:
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
line   25:
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
line   25:
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
```

The function in question is: https://github.com/xolox/vim-misc/blob/3e6b8fb6f03f13434543ce1f5d24f6a5d3f34f0b/autoload/xolox/misc/path.vim#L9-L47. Which actually seems well-formed.

Since I can't figure out what's wrong with the function that nvim thinks, I don't know how to reproduce this with nvim -u NORC. I noticed this problem because I just rebuilt nvim, my last rebuild was from about ~2 weeks ago. While bisecting, the majority of commits fail to build with this message:

```
/home/aktau/github/neovim/neovim/src/nvim/eval.c:7984:35: error: macro "strncmp" requires 3 arguments, but only 2 given
     if (strncmp(s, S_LEN("silent")) == 0) {
                                   ^
```

(BTW, I thought strncmp(3) wasn't a macro.)

At the point in the tree where this particular error has been fixed (29bad04f I think), another error comes into play: 

```
/home/aktau/github/neovim/neovim/src/nvim/eval.c: At top
level:
cc1: warning: unrecognized command line option "-Wno-static-in-inline"
[enabled by default]
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/ui.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/ui_bridge.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/undo.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/version.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/window.c.o
[100%] Linking C executable ../../bin/nvim
CMakeFiles/nvim.dir/eval.c.o: In function `f_getcompletion':
/home/aktau/github/neovim/neovim/src/nvim/eval.c:9465:
undefined reference to `get_tv_string'
/home/aktau/github/neovim/neovim/src/nvim/eval.c:9466:
undefined reference to `get_tv_string'
collect2: error: ld returned 1 exit status
make[3]: *** [bin/nvim] Error 1
make[3]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make[2]: *** [src/nvim/CMakeFiles/nvim.dir/all] Error 2
make[2]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make[1]: *** [all] Error 2
make[1]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make: *** [nvim] Error 2
```

Which makes the bisect take a very long time (lots of `git bisect skip`). And (I think) loop back upon itself, making it impossible to find the commit that caused this. I suspect @ZyX-I might know which one is the exact culprit.

After a couple of these loops, I got tired of it and saved the git bisect state:

```
05b74399  2017-03-29  Justin M. Keyes                                                                             build: remove unused get_preproc_output() call
2846d508  2017-03-29  lonerover                                                                                   vim-patch:7.4.2276 (#6393)
46efe144  2017-03-28  ZyX                                                                                         functests: Try sleeping a bit more
b9603218  2017-03-28  ZyX                                                                                         eval/executor: Fix check-single-includes
a1d590a0  2017-03-28  ZyX                              refs/bisect/bad                                            *: Use const char * in set_one_cmd_context
114eaa15  2017-03-28  ZyX                              refs/bisect/skip-114eaa15f009cbd3e3deb177a2a67affa430fbb8  eval/typval,api/buffer: Fix review comments
58e34e8d  2017-03-26  ZyX                                                                                         eval/typval: Allow NULL dict as tv_dict_get_callback() argument
f4256243  2017-03-25  ZyX                              refs/bisect/skip-f4256243dbdbd2858be716f9f52763bd9a76be4d  eval: Fix -Werror=unitialized from QB
29bad04f  2017-03-25  ZyX                              refs/bisect/skip-29bad04f9e2e70b7c1c198ca0d86d23282bee8a1  eval: Do not supply S_LEN to strncmp
8daf756f  2017-03-25  ZyX                              refs/bisect/skip-8daf756fb6f92bdeb39e473b34364afd5270dd99  unittests: Fix linter errors
7826ee1c  2017-03-25  ZyX                                                                                         unittests: Add tv_get_string* tests
e08b27ba  2017-03-25  ZyX                              HEAD                                                       unittests: Add tv_get number tests
4536c064  2017-03-25  ZyX                                                                                         unittests: Move tv_dict_add* tests to a proper describe() block
49195063  2017-03-25  ZyX                                                                                         unittests: Add tv_check… tests
389274be  2017-03-25  ZyX                                                                                         unittests: Add tv_equal() tests
630ff33d  2017-03-25  ZyX                                                                                         unittests: Test locks section
ed4948a9  2017-03-13  ZyX                                                                                         unittests: Test tv_copy()
f0bbd1e8  2017-03-13  ZyX                                                                                         unittests: Add tests for tv_clear()
e43de6bb  2017-03-13  ZyX                                                                                         unittests: Add test for tv_dict_set_keys_readonly
368a61c5  2017-03-13  ZyX                                                                                         unittests: Add tv_dict_copy tests
218fa1d8  2017-03-13  ZyX                              refs/bisect/skip-218fa1d8069c1f1d8c5154b1abc2b686adc0c742  charset: Remove useless condition from vim_iswordc_tab
8b9a1fbf  2017-03-13  ZyX                              refs/bisect/skip-8b9a1fbf7a630b68b1428a39f25e1fa38fe0cc9f  unittests: Add tests for tv_dict_extend
fa852e7c  2017-03-13  ZyX                              refs/bisect/skip-fa852e7cdc365b6fcd39d677f4067963274c44c3  eval: Fix extend() behaviour with NULL lists and dictionaries
4987850c  2017-03-13  ZyX                              refs/bisect/skip-4987850cacc9f0c5674bc4cb8197a6fa0a343167  unittests: Add tv_dict_clear tests
43e9fad1  2017-03-13  ZyX                              refs/bisect/skip-43e9fad1c8835a3136f4db53c82608e034df2a5e  eval: Use tv_is_func in place of ==VAR_FUNC                                                                                                                                                                                                                                                                                                        ==VAR_PARTIAL
270a3889  2017-03-13  ZyX                              refs/bisect/skip-270a3889af024485fa7b63f34c4dd3f92f6e0f98  unittests: Add tv_dict_add* unit tests
bc87d23c  2017-03-12  ZyX                              refs/bisect/skip-bc87d23c28c10c70b4addaf18ae16bcbe0682c8a  unittests: Add tests for dictionary indexing
b222453c  2017-03-12  ZyX                              refs/bisect/skip-b222453c95dbe466b79abd578642fa6038770a55  eval/typval: Refactor errors a bit: use emsgf always
5ce62432  2017-03-12  ZyX                              refs/bisect/skip-5ce624324136408a3f412d6c2ff437455f1e3344  unittests: Enable tv_list_join tests back
52e226ff  2017-03-05  ZyX                              refs/bisect/skip-52e226ff74f49a02f884247b61d1865fa14e810d  unittests: Disable tv_list_join test on Mac OS only
f830243f  2017-03-05  ZyX                              refs/bisect/skip-f830243ff7b78b9bb41e72fd602b99c2893b5437  mbyte: Include os_defs.h in mbyte.h
3bf87a5a  2017-03-05  ZyX                                                                                         eval: Do not use S_LEN as snprintf argument
38dd81c1  2017-03-05  ZyX                              refs/bisect/skip-38dd81c136c2edbda66614622a4747bf1785af94  eval/typval: Fix SEGV in test_alot.vim test
faddd83d  2017-03-05  ZyX                                                                                         eval: Fix SEGV in test49
6c622ed0  2017-03-05  ZyX                              refs/bisect/skip-6c622ed08bb56f08f4cc4dbf6251220ccf7ba2ba  unittests: Add tv_dict_item_{add,remove} tests
ffaf7c75  2017-03-05  ZyX                              refs/bisect/skip-ffaf7c7521399826d9a32b12a1180bcd162f88be  unittests: Add tv_dict_item_{alloc,free} tests
cdb1aa3e  2017-03-05  ZyX                              refs/bisect/skip-cdb1aa3e47cb0ec19d2ae597c1d21b7e892d0d7e  eval: Fix len argument to xstrlcat
2c8ad276  2017-03-05  ZyX                              refs/bisect/skip-2c8ad276523489223d2d804075a19367280b0f10  ascii: Readd DEL_STR define
c6c48e86  2017-03-04  ZyX                              refs/bisect/skip-c6c48e8672ab6adbf78453e2b133fe3bd2ee1eb5  syntax: Fix linter error
78a0de2c  2017-03-04  ZyX                              refs/bisect/skip-78a0de2c1b722d5445dfc44cc0635e5ce4bcd23e  eval/typval: Fix -Werror=return-type
4c3be98d  2017-03-04  ZyX                              refs/bisect/skip-4c3be98db9b24b7d5b8dc2cf55574e778b7ad137  unittests: Add tv_dict_watcher_{add,remove} tests
14017466  2017-03-04  ZyX                              refs/bisect/skip-140174669e359ae7147288366dd94952955980cc  unittests: Run tv_list_join tests in case it stopped failing
6aa6e500  2017-03-04  ZyX                              refs/bisect/skip-6aa6e5007596f1eaa23c45376e3a72d908c6688e  eval: Fix linter errors
506b9389  2017-02-14  ZyX                              refs/bisect/skip-506b938947b7083fc8bef20eabc08ed033298add  *: Make some more things const and with length
3025431c  2016-12-04  ZyX                              refs/bisect/skip-3025431c81daac873e69a71aee695ebfd00504f7  eval: Make sure that v:_null_dict does not crash dictwatcher*()
a56f2d27  2016-12-04  ZyX                              refs/bisect/skip-a56f2d27e3c09aaae00a58a70652ac5db3287dee  eval: Make dictionary watchers work with empty keys
1e3e302d  2016-12-04  ZyX                              refs/bisect/skip-1e3e302dc2bdced563ecd2c3fb101af31f72b3df  eval: Move part of dictwatcher* functions to eval/typval
52396162  2016-11-05  ZyX                              refs/bisect/skip-5239616297182601a5d244a482e6cef307e901f1  functests: Fix buf_functions test on Windows
4bcee963  2016-11-05  ZyX                              refs/bisect/skip-4bcee963471abd939bb9edd1709418e30be7290f  *: Fix some Windows-specific warnings
9ed9af7e  2016-11-05  ZyX                              refs/bisect/skip-9ed9af7e11e3a707f65abfeb1d02b029e39241ea  eval/typval: More `const` qualifiers in `tv_dict*` function signatures
56e51033  2016-11-05  ZyX                              refs/bisect/skip-56e51033abf00d66e9c6f9412e8f57c9a24b86ae  unittests: Add tests for tv_list_idx_of_item
e5edf07e  2016-11-04  ZyX                              refs/bisect/skip-e5edf07ec44f8d147d7482cae2997be62c30373f  unittests: Add tests for tv_list_find*() functions
b3672ae2  2016-10-02  ZyX                              refs/bisect/skip-b3672ae2fced4715963442d2e19048f8fadbe0b8  eval/typval: Add tv_list_equal() tests, compare NULL lists equal
4f9e7844  2016-09-25  ZyX                              refs/bisect/skip-4f9e7844272335943c075b9444934f77d3cba234  unittests: Test tv_list_join()
cf45c7bb  2016-09-25  ZyX                              refs/bisect/skip-cf45c7bb059dafeb882133273f77042cb06b37e4  unittests: Fix tests crash
7ceebacb  2016-09-24  ZyX                                                                                         eval/typval,tests: Fix extending list with itself, add tests
56e4c2f6  2016-09-19  ZyX                                                                                         unittests: Test tv_list_concat()
f80a0046  2016-09-18  ZyX                              refs/bisect/skip-f80a00469fdba8a3dec0edae30c911d050485055  eval/typval: Make tv_list_concat handle NULL lists correctly
82e6cac5  2016-09-18  ZyX                              refs/bisect/skip-82e6cac5f9d437da5aeb8c4cf51945cfeb5922ff  functests: Add null_spec.lua from #4615
9898f36a  2016-09-17  ZyX                              refs/bisect/skip-9898f36aa306d2a7f53fbe08c64048260992d3bb  unittests: Test tv_list_copy
9b8beaff  2016-09-17  ZyX                                                                                         unittests: Add tests for tv_list_insert*()/…append*() functions
be360d88  2016-09-14  ZyX                              refs/bisect/skip-be360d88414fa3b874262f3ed3a2bba5926a751e  unittests: Add tests for tv_list_insert()
d2639e1e  2016-09-14  ZyX                                                                                         unittests: Add tests for list watchers and list alloc/free/unref
a3941671  2017-03-12  ZyX                              refs/bisect/skip-a394167177390a66c275d32e3862c82c546970ff  unittests: Test tv_list_item_\* functions
a32db8ed  2016-09-11  ZyX                              refs/bisect/skip-a32db8ed19c35847373d4a7fd56d7797a5a26897  eval/typval: Add missing includes, also add a script to find them
2ad4fba4  2016-09-11  ZyX                              refs/bisect/skip-2ad4fba46db26d0724106b194f7cb628fb9b02e8  eval: Move copy_tv to eval/typval
8b0fa64e  2016-09-10  ZyX                              refs/bisect/skip-8b0fa64ed3dbd4376cbfa195526ec97d8c944a97  message: Remove incorrect assertion
40feac6e  2016-09-10  ZyX                                                                                         message: Revise maxlen argument in msg_puts_attr_len
3a3816c9  2016-09-04  ZyX                              refs/bisect/skip-3a3816c990362e1ad8321e6b4748a33c94d1d797  cmake: Use CMAKE_CURRENT_LIST_DIR and remove vars used only once
86fc4580  2016-09-04  ZyX                              refs/bisect/skip-86fc4580b83f20211d7f85566e84a0e1dad0a136  eval: Fix max_min functions
c4fe656f  2016-09-04  ZyX                              refs/bisect/skip-c4fe656fef5ade20da4861a2f1e5d0f776b0df8f  typval.h: Allow non-var expressions in TV_DICT_ITER first argument
6cc3d59e  2016-09-04  ZyX                              refs/bisect/skip-6cc3d59ec8e4d6f32c8c3d9755c625e32512b8e2  misc1: Refactor ask_yesno()
31a3158d  2016-09-04  ZyX                              refs/bisect/skip-31a3158d0b574385186ab55c074edf85356d1d6c  eval: Make sort always stable
5df35297  2016-09-04  ZyX                              refs/bisect/skip-5df35297f832b3247c18253c916be6066c603739  eval: Remove eval_expr() completely
c8e63a8d  2016-09-04  ZyX                              refs/bisect/skip-c8e63a8db84e9d9f7bd855085a87d93631504fc7  eval: Move remaining get_tv_string* functions to eval/typval.c
50ebd1df  2016-08-28  ZyX                              refs/bisect/skip-50ebd1dff5c4e995c4f7e7980870e43d9defabc6  eval: Move free_tv to eval/typval.h, remove most of its usages
233b0c93  2016-08-28  ZyX                                                                                         eval: Move get_tv_number[_chk] to eval/typval.c
1b3e13da  2016-08-28  ZyX                              refs/bisect/skip-1b3e13da5be48739ac4291208e421b68e607cc0f  eval: Refactor get_tv_lnum_buf
7ee5cc74  2016-08-28  ZyX                              refs/bisect/skip-7ee5cc7429017a4a08a8b62b628bea156fea0008  eval: Move get_tv_lnum and get_tv_float to eval/typval.h
949f09bd  2016-08-21  ZyX                                                                                         eval: Move get_tv_string_buf() to eval/typval.c
28dafe3f  2016-08-21  ZyX                              refs/bisect/skip-28dafe3ff0b0dc082fb62b2251fd64a167ce7188  eval,*: Move get_tv_string to typval.c
5cdf7177  2016-08-21  ZyX                                                                                         eval: Move get_float_arg to typval.h
2dcfc439  2016-08-20  ZyX                                                                                         eval: Split and move dict_add_nr_str to typval.c
210342d7  2017-01-11  ZyX                              refs/bisect/good-210342d795a13e88579d3e55ae931463d16f241d  eval: Move dict_add_list and dict_add_dict to typval.c
983a5532  2016-08-20  ZyX                              refs/bisect/good-983a5532ca17060508ea308a9b7069dae4facb59  eval: Move dict_set_keys_readonly to typval.c
ecff8387  2016-08-20  ZyX                                                                                         eval: Move get_dict_callback to typval.c
54bd2e8b  2016-08-20  ZyX                                                                                         eval: Make setmatches() return -1 in case of some failures
e18a5783  2016-08-20  ZyX                              refs/bisect/good-e18a5783080f7c94f408ec5f53dedffdb69789e1  *: Move some dictionary functions to typval.h and use char*
50a48f2a  2016-07-30  ZyX                              refs/bisect/good-50a48f2a0ecf7f767df961f7f5060505cf28e331  functests: Add tests for some *buf* functions
fb146e80  2016-07-26  ZyX                                                                                         eval: Split eval.c into smaller files
18e7d552  2017-03-29  Justin M. Keyes                                                                             terminal.c:redraw(): Avoid invalid cursor col (#6379)
e86042ae  2017-03-29  lonerover                                                                                   vim-patch:7.4.2343 and mark NA patches (#6384)
b9e7ab14  2017-03-28  Samuel Catherasoo                                                                           refactor/single-include: charset.h (#6385)
```

The ones with `skip-` refs don't compile for some of the reasons listed above. The culprit is somewhere in there, but I'd need to make patches for the compilation issues and apply them during bisection. Keeping this bug to pick it up when I get time.

**Comment by ZyX-I** [comment]
The problem is that I did not intentionally touch anything which should cause E171, so I really do not know. And with git bisect you may understand why mercurial uses immutable history: I did not really make most of those commits, this is rebase. And while commit state was consistent when *I* was making a commit, after rebase it is no longer guaranteed. You have a bit better situation then it could possibly be because first few rebases I was checking whether tests are fine after solving conflicts.

Basically current development model keeps clean history, but is by no means good for any more or less big PR. E.g. even #6236 has inconsistent state in the middle of the PR: “Fix comments and cleanup tests” 3ea644c7c9e5c94d327fa6cf0b78470d2d0d2698 commit uses `rettv_dict_alloc`, but there is no such function and you can no longer bisect. And that is not a big PR.

**Comment by justinmk** [comment]
> Basically current development model keeps clean history, but is by no means good for any more or less big PR.

Would it help if we changed the recommendation to:

1. Rebase and squash the feature branch on its _current_ origin.
2. Then _merge_ to upstream HEAD, don't rebase.

**Comment by ZyX-I** [comment]
I would suggest using `git ri` for interactive rebase to clean up `fixup!`s and merging upstream:

```
	ri = !zsh -c 'git rebase --interactive $(git merge-base --is-ancestor ${1:-master} ${2:-HEAD} && git rev-parse ${1:-master} || git merge-base ${1:-master} ${2:-HEAD}) ${@[3,-1]}' -
```

(usage: `git ri` if PR to master, `git ri X` if PR is supposed to be merged into X; never actually used $2).

Note: if I am not mistaking, `git ri` after merging will no longer allow altering post-merge commits (unless you use second argument). Though think this is a feature.

Alias is using zsh, but there are no zsh-ism apart from `${@[3, -1]}` and the fact that I did not bother with quoting (everything, but `$()` will appear in one piece in zsh), easy to fix.

If I am not mistaking, this is exactly what you mean by “rebase and squash the feature branch on its current origin”.

**Comment by ZyX-I** [comment]
Additionally, it is fine to rebase the *single* commit: there will be no inconsistent state then.

**Comment by ZyX-I** [comment]
Wondering whether my alias may just use `--fork-point`. Did not mention that, or is it relatively recent addition?

**Comment by ZyX-I** [comment]
No, this is the default behaviour. Used to determine what to rebase, not the destination.

**Comment by ZyX-I** [comment]
After translating this to POSIX shell it became even more lengthy:

   	ri = "!sh -c 't=\"${1:-master}\"; s=\"${2:-HEAD}\" ; [ $# -gt 0 ] && shift ; [ $# -gt 0 ] && shift ; git rebase --interactive \"$(git merge-base --is-ancestor \"$t\" \"$s\" && git rev-parse \"$t\" || git merge-base \"$t\" \"$s\")\" \"$@\"' -"

(tested with `dash` in place of `sh`: my `sh` is `bash` symlink). It looks like on `#git` people either do not understand the request or have no idea how to implement it without such big script.

**Comment by aktau** [comment]
Another update: I can reproduce it with a running nvim, but not reliably (yea...):

```
:echo xolox#misc#path#which('awk')
Error detected while processing function xolox#misc#path#which:                     
line   25:                                                                          
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
line   34:                                                                          
E580: :endif without :if:       endif                                               
line   25:                                                                          
E171: Missing :endif:         endfor                                                
line   34:                                                                          
E580: :endif without :if:       endif                                               
line   36:                                                                          
E588: :endfor without :for:   endfor                                                
['/home/aktau/.linuxbrew/bin/awk', '/usr/bin/awk'] 

:echo xolox#misc#path#which('awk')
['/home/aktau/.linuxbrew/bin/awk', '/usr/bin/awk'] 
```

Re-running the same call afterwards no longer provokes the error. I thought eval.c always re-evaluated code? How is this possible?

Note that it doesn't appear required that the function returns multiple paths. The error I get is probably based on asking for `which ctags`:

```
:echo xolox#misc#path#which('ctags')

Error detected while processing function xolox#misc#path#which:                     
line   25:                                                                          
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
E171: Missing :endif:         endfor                                                
line   34:                                                                          
E580: :endif without :if:       endif                                               
line   25:                                                                          
E171: Missing :endif:         endfor                                                
line   34:                                                                          
E580: :endif without :if:       endif                                               
line   36:                                                                          
E588: :endfor without :for:   endfor                                                
['/usr/bin/ctags']  
```

It's interesting to note that despite nvim's complaints, it still returns the correct result (`/usr/bin/ctags`). That's likely why I've found no negative effects from this apart from logspam.

I haven't been able to reproduce this from my macOS machine, (though there I don't have ctags installed).

**Comment by jamessan** [comment]
Have you reproducing with an ASAN build or running through valgrind?

**Comment by aktau** [comment]
No. Do you have an invocation handy that would give me an ASan build? I have clang-3.8 here locally.

**Comment by jamessan** [comment]
```sh
$ mkdir build
$ cd build
$ env CC=clang cmake -DCLANG_ASAN_UBSAN=ON ..
$ make
```

**Comment by aktau** [comment]
Hmmm, it doesn't find the packages apparently.

```
$  rm -rf build .deps ; ( mkcd build && brewless env CC=clang-3.8 CXX=clang++-3.8 cmake -DCLANG_ASAN_UBSAN=ON -DCMAKE_INSTALL_PREFIX:PATH=$HOME/neovim .. ) && make
-- The C compiler identification is Clang 3.8.0
-- The CXX compiler identification is Clang 3.8.0
-- Check for working C compiler: /usr/bin/clang-3.8
-- Check for working C compiler: /usr/bin/clang-3.8 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-3.8
-- Check for working CXX compiler: /usr/bin/clang++-3.8 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- CMAKE_BUILD_TYPE not given, defaulting to 'Dev'.
-- Replacing -O3 in CMAKE_C_FLAGS_RELEASE with -O2.
-- Performing Test HAS_ACCEPTABLE_FORTIFY
-- Performing Test HAS_ACCEPTABLE_FORTIFY - Success
...
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.26") 
...
CMake Error at /usr/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find LibUV (missing: LIBUV_LIBRARY LIBUV_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLibUV.cmake:105 (find_package_handle_standard_args)
  CMakeLists.txt:307 (find_package)
```

brewless is (similar result with and without brewless):

```
brewless () {
	(
		env | grep linuxbrew | sed -r -e 's@([=:])[^:]*.linuxbrew[^:]*:@\1@' | while read i
		do
			export $i
		done
		"$@"
	)
}
```

My normal command is:

```
 make distclean && make -j12 CMAKE_BUILD_TYPE=RelWithDebInfo CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH=$HOME/neovim" install
```

Which works. So I need (or someone more experienced) needs to figure out what the difference is.

**Comment by jamessan** [comment]
Just add `-DCLANG_ASAN_UBSAN=ON` to your `CMAKE_EXTRA_FLAGS`.

**Comment by aktau** [comment]
Yep, it crashes. Though I get no symbolization:

```
$ make distclean && make CC=clang-3.8 CXX=clang++-3.8 CMAKE_EXTRA_FLAGS="-DCLANG_ASAN_UBSAN=ON -DCMAKE_INSTALL_PREFIX:PATH=$HOME/neovim" install
...
$ nvim
=================================================================
==1246==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60400010f2c0 at pc 0x000000968320 bp 0x7ffd3d9b29c0 sp 0x7ffd3d9b29b8
READ of size 1 at 0x60400010f2c0 thread T0
    #0 0x96831f  (/home/aktau/neovim/bin/nvim+0x96831f)
    #1 0x62e83b  (/home/aktau/neovim/bin/nvim+0x62e83b)
    #2 0x5ea5e8  (/home/aktau/neovim/bin/nvim+0x5ea5e8)
    #3 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #4 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #5 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #6 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #7 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #8 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #9 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #10 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #11 0x5ec0d5  (/home/aktau/neovim/bin/nvim+0x5ec0d5)
    #12 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #13 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #14 0x5fd080  (/home/aktau/neovim/bin/nvim+0x5fd080)
    #15 0x5ea51d  (/home/aktau/neovim/bin/nvim+0x5ea51d)
    #16 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #17 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #18 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #19 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #20 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #21 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #22 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #23 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #24 0x5ec0d5  (/home/aktau/neovim/bin/nvim+0x5ec0d5)
    #25 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #26 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #27 0x5fd080  (/home/aktau/neovim/bin/nvim+0x5fd080)
    #28 0x5ea51d  (/home/aktau/neovim/bin/nvim+0x5ea51d)
    #29 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #30 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #31 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #32 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #33 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #34 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #35 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #36 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #37 0x5ede1f  (/home/aktau/neovim/bin/nvim+0x5ede1f)
    #38 0x735834  (/home/aktau/neovim/bin/nvim+0x735834)
    #39 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #40 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #41 0x5fd080  (/home/aktau/neovim/bin/nvim+0x5fd080)
    #42 0x5ea51d  (/home/aktau/neovim/bin/nvim+0x5ea51d)
    #43 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #44 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #45 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #46 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #47 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #48 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #49 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #50 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #51 0x5e69fb  (/home/aktau/neovim/bin/nvim+0x5e69fb)
    #52 0x733fc0  (/home/aktau/neovim/bin/nvim+0x733fc0)
    #53 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #54 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #55 0x6ec20a  (/home/aktau/neovim/bin/nvim+0x6ec20a)
    #56 0x6e961c  (/home/aktau/neovim/bin/nvim+0x6e961c)
    #57 0x6e9c86  (/home/aktau/neovim/bin/nvim+0x6e9c86)
    #58 0x6e94a2  (/home/aktau/neovim/bin/nvim+0x6e94a2)
    #59 0x6e924f  (/home/aktau/neovim/bin/nvim+0x6e924f)
    #60 0x7ff9fd  (/home/aktau/neovim/bin/nvim+0x7ff9fd)
    #61 0x7fc6fe2a6f44  (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #62 0x447fb5  (/home/aktau/neovim/bin/nvim+0x447fb5)

0x60400010f2c0 is located 0 bytes to the right of 48-byte region [0x60400010f290,0x60400010f2c0)
allocated by thread T0 here:
    #0 0x4f2318  (/home/aktau/neovim/bin/nvim+0x4f2318)
    #1 0x8508f9  (/home/aktau/neovim/bin/nvim+0x8508f9)
    #2 0x62e819  (/home/aktau/neovim/bin/nvim+0x62e819)
    #3 0x5ea5e8  (/home/aktau/neovim/bin/nvim+0x5ea5e8)
    #4 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #5 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #6 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #7 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #8 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #9 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #10 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #11 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #12 0x5ec0d5  (/home/aktau/neovim/bin/nvim+0x5ec0d5)
    #13 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #14 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #15 0x5fd080  (/home/aktau/neovim/bin/nvim+0x5fd080)
    #16 0x5ea51d  (/home/aktau/neovim/bin/nvim+0x5ea51d)
    #17 0x5f1da7  (/home/aktau/neovim/bin/nvim+0x5f1da7)
    #18 0x61feb2  (/home/aktau/neovim/bin/nvim+0x61feb2)
    #19 0x61c1eb  (/home/aktau/neovim/bin/nvim+0x61c1eb)
    #20 0x61b4d3  (/home/aktau/neovim/bin/nvim+0x61b4d3)
    #21 0x619fd5  (/home/aktau/neovim/bin/nvim+0x619fd5)
    #22 0x619a28  (/home/aktau/neovim/bin/nvim+0x619a28)
    #23 0x5e79c2  (/home/aktau/neovim/bin/nvim+0x5e79c2)
    #24 0x5e7334  (/home/aktau/neovim/bin/nvim+0x5e7334)
    #25 0x5ec0d5  (/home/aktau/neovim/bin/nvim+0x5ec0d5)
    #26 0x6ffb14  (/home/aktau/neovim/bin/nvim+0x6ffb14)
    #27 0x6f39a6  (/home/aktau/neovim/bin/nvim+0x6f39a6)
    #28 0x5fd080  (/home/aktau/neovim/bin/nvim+0x5fd080)
    #29 0x5ea51d  (/home/aktau/neovim/bin/nvim+0x5ea51d)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/aktau/neovim/bin/nvim+0x96831f) 
Shadow bytes around the buggy address:
  0x0c0880019e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c0880019e50: fa fa 00 00 00 00 00 00[fa]fa 00 00 00 00 03 fa
  0x0c0880019e60: fa fa fd fd fd fd fd fa fa fa 00 00 00 00 03 fa
  0x0c0880019e70: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
  0x0c0880019e80: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fa
  0x0c0880019e90: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
  0x0c0880019ea0: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1246==ABORTING
```

**Comment by aktau** [comment]
Got one (removing the old one later to not make the issue too long):

```
$ ASAN_OPTIONS=symbolize=1 ASAN_SYMBOLIZER_PATH=/usr/lib/llvm-3.8/bin/llvm-symbolizer nvim
=================================================================
==2029==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60400010f2c0 at pc 0x000000968320 bp 0x7fffd9f5fb80 sp 0x7fffd9f5fb78
READ of size 1 at 0x60400010f2c0 thread T0
    #0 0x96831f in path_tail /home/aktau/github/neovim/neovim/src/nvim/path.c:97:10
    #1 0x62e83b in f_resolve /home/aktau/github/neovim/neovim/src/nvim/eval.c:13424:18
    #2 0x5ea5e8 in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6385:11
    #3 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #4 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #5 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #6 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #7 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #8 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #9 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #10 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #11 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #12 0x5ec0d5 in ex_let /home/aktau/github/neovim/neovim/src/nvim/eval.c:1495:9
    #13 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #14 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #15 0x5fd080 in call_user_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:21033:3
    #16 0x5ea51d in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6370:11
    #17 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #18 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #19 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #20 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #21 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #22 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #23 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #24 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #25 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #26 0x5ec0d5 in ex_let /home/aktau/github/neovim/neovim/src/nvim/eval.c:1495:9
    #27 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #28 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #29 0x5fd080 in call_user_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:21033:3
    #30 0x5ea51d in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6370:11
    #31 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #32 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #33 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #34 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #35 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #36 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #37 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #38 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #39 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #40 0x5ede1f in eval_for_line /home/aktau/github/neovim/neovim/src/nvim/eval.c:2526:7
    #41 0x735834 in ex_while /home/aktau/github/neovim/neovim/src/nvim/ex_eval.c:979:14
    #42 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #43 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #44 0x5fd080 in call_user_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:21033:3
    #45 0x5ea51d in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6370:11
    #46 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #47 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #48 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #49 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #50 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #51 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #52 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #53 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #54 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #55 0x5e69fb in eval_to_bool /home/aktau/github/neovim/neovim/src/nvim/eval.c:955:7
    #56 0x733fc0 in ex_if /home/aktau/github/neovim/neovim/src/nvim/ex_eval.c:806:14
    #57 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #58 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #59 0x6ec20a in do_source /home/aktau/github/neovim/neovim/src/nvim/ex_cmds2.c:2932:3
    #60 0x6e961c in source_callback /home/aktau/github/neovim/neovim/src/nvim/ex_cmds2.c:2315:9
    #61 0x6e9c86 in do_in_path /home/aktau/github/neovim/neovim/src/nvim/ex_cmds2.c:2403:15
    #62 0x6e94a2 in do_in_runtimepath /home/aktau/github/neovim/neovim/src/nvim/ex_cmds2.c:2447:12
    #63 0x6e924f in source_runtime /home/aktau/github/neovim/neovim/src/nvim/ex_cmds2.c:2325:10
    #64 0x7ff9fd in load_plugins /home/aktau/github/neovim/neovim/src/nvim/main.c:1281:5
    #65 0x7ff9fd in main /home/aktau/github/neovim/neovim/src/nvim/main.c:353
    #66 0x7f643a455f44 in __libc_start_main /build/eglibc-MjiXCM/eglibc-2.19/csu/libc-start.c:287
    #67 0x447fb5 in _start (/home/aktau/neovim/bin/nvim+0x447fb5)

0x60400010f2c0 is located 0 bytes to the right of 48-byte region [0x60400010f290,0x60400010f2c0)
allocated by thread T0 here:
    #0 0x4f2318 in malloc (/home/aktau/neovim/bin/nvim+0x4f2318)
    #1 0x8508f9 in try_malloc /home/aktau/github/neovim/neovim/src/nvim/memory.c:84:15
    #2 0x8508f9 in xmalloc /home/aktau/github/neovim/neovim/src/nvim/memory.c:118
    #3 0x62e819 in f_resolve /home/aktau/github/neovim/neovim/src/nvim/eval.c:13422:17
    #4 0x5ea5e8 in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6385:11
    #5 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #6 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #7 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #8 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #9 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #10 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #11 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #12 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #13 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #14 0x5ec0d5 in ex_let /home/aktau/github/neovim/neovim/src/nvim/eval.c:1495:9
    #15 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #16 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #17 0x5fd080 in call_user_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:21033:3
    #18 0x5ea51d in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6370:11
    #19 0x5f1da7 in get_func_tv /home/aktau/github/neovim/neovim/src/nvim/eval.c:6125:11
    #20 0x61feb2 in eval7 /home/aktau/github/neovim/neovim/src/nvim/eval.c:4241:15
    #21 0x61c1eb in eval6 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3944:7
    #22 0x61b4d3 in eval5 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3792:7
    #23 0x619fd5 in eval4 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3529:7
    #24 0x619a28 in eval3 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3447:7
    #25 0x5e79c2 in eval2 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3378:7
    #26 0x5e79c2 in eval1 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3305
    #27 0x5e7334 in eval0 /home/aktau/github/neovim/neovim/src/nvim/eval.c:3265:9
    #28 0x5ec0d5 in ex_let /home/aktau/github/neovim/neovim/src/nvim/eval.c:1495:9
    #29 0x6ffb14 in do_one_cmd /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:2209:5
    #30 0x6f39a6 in do_cmdline /home/aktau/github/neovim/neovim/src/nvim/ex_docmd.c:603:20
    #31 0x5fd080 in call_user_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:21033:3
    #32 0x5ea51d in call_func /home/aktau/github/neovim/neovim/src/nvim/eval.c:6370:11

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/aktau/github/neovim/neovim/src/nvim/path.c:97:10 in path_tail
Shadow bytes around the buggy address:
  0x0c0880019e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0880019e40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c0880019e50: fa fa 00 00 00 00 00 00[fa]fa 00 00 00 00 03 fa
  0x0c0880019e60: fa fa fd fd fd fd fd fa fa fa 00 00 00 00 03 fa
  0x0c0880019e70: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
  0x0c0880019e80: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fa
  0x0c0880019e90: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
  0x0c0880019ea0: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==2029==ABORTING
```

**Comment by aktau** [comment]
It would seem the error is here:

```
/// Gets the tail (i.e., the filename segment) of a path `fname`.
///
/// @return pointer just past the last path separator (empty string, if fname
///         ends in a slash), or empty string if fname is NULL.
char_u *path_tail(const char_u *fname)
  FUNC_ATTR_NONNULL_RET
{
  if (fname == NULL) {
    return (char_u *)"";
  }

  const char_u *tail = get_past_head(fname);
  const char_u *p = tail;
  // Find last part of path.
  while (*p != NUL) {  // NULL dereference.
    if (vim_ispathsep_nocolon(*p)) {
      tail = p + 1;
    }
    mb_ptr_adv(p);
  }
  return (char_u *)tail;
}

/// Get a pointer to one character past the head of a path name.
/// Unix: after "/"; Win: after "c:\"
/// If there is no head, path is returned.
char_u *get_past_head(const char_u *path)
{
  const char_u *retval = path;

#ifdef WIN32
  // May skip "c:"
  if (isalpha(path[0]) && path[1] == ':') {
    retval = path + 2;
  }
#endif

  while (vim_ispathsep(*retval)) {
    ++retval;
  }

  return (char_u *)retval;
}

/*
 * Return TRUE if 'c' is a path separator.
 * Note that for MS-Windows this includes the colon.
 */
int vim_ispathsep(int c)
{
#ifdef UNIX
  return c == '/';          // Unix has ':' inside file names
#else
# ifdef BACKSLASH_IN_FILENAME
  return c == ':' || c == '/' || c == '\\';
# else
  return c == ':' || c == '/';
# endif
#endif
}
```

I don't really think it's the `get_past_head` thing, but perhaps the `mb_ptr_adv` implementation changed recently  (perhaps #6335). Any thoughts @bfredl?

**Comment by aktau** [comment]
Scratch that, it's more likely to be the allocating function `f_resolve()`. Changed by @ZyX-I's  28dafe3ff0b0dc082fb62  (`eval,*: Move get_tv_string to typval.c`) or somewhere in that vicinity. It matches the good/bad results I was able to get out of `git-bisect`.

**Comment by oni-link** [comment]
Could be that `cpy` is not yet a C-string. Can you try this patch?
```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 7ab07fe6a..3924d8c79 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -13421,6 +13421,7 @@ static void f_resolve(typval_T *argvars, typval_T *rettv, FunPtr fptr)
           const size_t buf_len = strlen(buf);
           cpy = xmalloc(p_len + buf_len + 1);
           memcpy(cpy, p, p_len);
+          cpy[p_len] = 0;
           memcpy(path_tail((char_u *)cpy), buf, buf_len + 1);
           xfree(p);
           p = cpy;
```

**Comment by aktau** [comment]
```
          // Symlink is relative to directory of argument.
          const size_t p_len = strlen(p);
          const size_t buf_len = strlen(buf);
          cpy = xmalloc(p_len + buf_len + 1);
          memcpy(cpy, p, p_len);
          memcpy(path_tail((char_u *)cpy), buf, buf_len + 1);
          xfree(p);
```

(There must be an invariant somewhere that `path_tail` does not return `""` somewhere, otherwise this code would otherwise be buggy too.)

This seems like it could use something simpler:

```
p = xrealloc(p, strlen(p) + strlen(buf) + 1); // Allocate maximum possible path length.
strlcpy(path_tail(p), buf, strlen(buf)+1); // Replace the filename.
```

Though I only tried to look at what the original bit of code was doing, not why it was wrong (some off by one?).

**Comment by aktau** [comment]
@oni-link link was right, as usual.

That said, wouldn't `strlcpy` have avoided this problem in the first place? Perhaps it's better to shorten this code than to add a hotpatch.

**Comment by oni-link** [comment]
I assume `(x)strlcpy()` was not used because of extra `strlen()` evaluations. I like your `xrealloc()` solution to shorten the code:
```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 7ab07fe6a..cecaab34a 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -13419,11 +13419,8 @@ static void f_resolve(typval_T *argvars, typval_T *rettv, FunPtr fptr)
           // Symlink is relative to directory of argument.
           const size_t p_len = strlen(p);
           const size_t buf_len = strlen(buf);
-          cpy = xmalloc(p_len + buf_len + 1);
-          memcpy(cpy, p, p_len);
-          memcpy(path_tail((char_u *)cpy), buf, buf_len + 1);
-          xfree(p);
-          p = cpy;
+          p = xrealloc(p, p_len + buf_len + 1);
+          memcpy(path_tail((char_u *)p), buf, buf_len + 1);
         } else {
           xfree(p);
           p = xstrdup(buf);
```

**Comment by aktau** [comment]
That looks good, if I don't get to it soon (maybe tonight), please anyone feel free to commit it.

**Comment by aktau** [comment]
On a related note, it could be a net win to refactor `path_tail` to take and return `char *`:

```
$ rg 'path_tail\(' | awk '/char_u \*/ || /char \*/{ print "better", $0 ; next} {print "worse", $0 }' | sort
better nvim/eval.c:    if (*path_tail((char_u *)dir) == NUL) {
better nvim/eval.c:          memcpy(path_tail((char_u *)p), buf, buf_len + 1);
better nvim/eval.c:          q = (char *)path_tail((char_u *)p);
better nvim/eval.c:        q = (char *)path_tail((char_u *)p);
better nvim/event/socket.c:      *path_tail((char_u *)watcher->addr) = NUL;
better nvim/fileio.c:      char * tail = (char *)path_tail(tempname);
better nvim/fileio.c:    if (p_fic && (STRCMP(path_tail((char_u *)from), path_tail((char_u *)to))
better nvim/fileio.c:  if (prepend_dot && *(e = (char *)path_tail((char_u *)retval)) != '.') {
better nvim/fileio.c:          sprintf((char *)path_tail(IObuff), "%d", i);
better nvim/file_search.c:      char_u *p =  path_tail(search_ctx->ffsc_fix_path);
better nvim/if_cscope.c:        path_tail((char_u *)csinfo[i].fname)
better nvim/main.c:  argv0 = (char *)path_tail((char_u *)argv[0]);
better nvim/main.c:        r = (char_u *)concat_fnames((char *)p, (char *)path_tail(alist_name(&GARGLIST[0])), TRUE);
better nvim/main.c:  set_vim_var_string(VV_PROGNAME, (char *)path_tail((char_u *)exename), -1);
better nvim/memline.c:          msg_puts((const char *)path_tail(files[i]));
better nvim/option.c:  char_u *p = (char_u *)invocation_path_tail(p_sh, &len);
better nvim/option.c:  return strstr((char *)path_tail(p_sh), "csh") != NULL;
better nvim/os/env.c:      char *vim_path_end = (char *)path_tail((char_u *)vim_path);
better nvim/os/env.c:        path_end = (char *)path_tail((char_u *)exe_name);
better nvim/os/env.c:    xstrlcpy((char *)dst, (char *)path_tail(src), dstlen);
better nvim/os_unix.c:  if (shell_style == STYLE_ECHO && strstr((char *)path_tail(p_sh),
better nvim/path.c:char_u *path_tail(const char_u *fname)
better nvim/path.c:      char_u *p = path_tail(curbuf->b_ffname);
better nvim/path.c:      char_u *tail = path_tail(fname);
better nvim/path.c:  char_u *tail = path_tail(fname);
better nvim/path.c:  char_u *tail = path_tail(str);
better nvim/path.c:const char_u *invocation_path_tail(const char_u *invocation, size_t *len)
better nvim/spellfile.c:    if (strstr((char *)path_tail(wfname), SPL_FNAME_ADD) != NULL)
better nvim/spellfile.c:    if (strstr((char *)path_tail(wfname), SPL_FNAME_ASCII) != NULL)
better nvim/spellfile.c:    lp->sl_add = strstr((char *)path_tail(fname), SPL_FNAME_ADD) != NULL;
better nvim/spellfile.c:                       && strstr((char *)path_tail(fname), ".ascii.") != NULL)
better nvim/undo.c:      char *const tail = (char *) path_tail((char_u *) undo_file_name);
worse nvim/buffer.c:        i_name = path_tail(curbuf->b_ffname);
worse nvim/buffer.c:        p = transstr(path_tail(curbuf->b_fname));
worse nvim/buffer.c:        str = path_tail(NameBuff);
worse nvim/eval.c:  tail = path_tail(*fnamep);
worse nvim/ex_cmds2.c:    p = path_tail(fname_exp);
worse nvim/ex_cmds.c:                e2 = vim_strrchr(path_tail(f2), '.');
worse nvim/ex_cmds.c:  fname = path_tail(curbuf->b_fname);
worse nvim/ex_cmds.c:    STRNCMP(invocation_path_tail(p_sh, NULL), "fish", 4) == 0;
worse nvim/ex_cmds.c:                t1 = path_tail(f1);
worse nvim/ex_docmd.c:      if ((s = vim_strrchr(result, '.')) != NULL && s >= path_tail(result))
worse nvim/ex_getln.c:  end = path_tail(xp->xp_pattern);
worse nvim/ex_getln.c:    s = path_tail(match);
worse nvim/ex_getln.c:    tail = path_tail(retval);
worse nvim/fileio.c:  char_u      *tail = path_tail(sfname);
worse nvim/fileio.c:    path_tail(vim_tempdir)[-1] = NUL;
worse nvim/fileio.c:  tail = path_tail(fname);
worse nvim/fileio.c:  tail = path_tail(sfname);
worse nvim/file_search.c:    size_t len = (size_t)(path_tail(rel_fname) - rel_fname);
worse nvim/file_search.c:          STRCPY(path_tail(NameBuff), ff_file_to_find);
worse nvim/memline.c:              if (fnamecmp(path_tail(buf->b_ffname),
worse nvim/memline.c:                           path_tail(b0.b0_fname)) != 0
worse nvim/memline.c:  tail = path_tail(fname);
worse nvim/memline.c:          tail = path_tail(fname_res);
worse nvim/memline.c:      tail = path_tail(tmp);
worse nvim/os_unix.c:    STRNCMP(invocation_path_tail(p_sh, NULL), "fish", 4) == 0;
worse nvim/quickfix.c:          fname = path_tail(fname);
worse nvim/quickfix.c:          STRLCPY(IObuff, path_tail(errbuf->b_fname), sizeof(IObuff));
worse nvim/spell.c:      p = vim_strchr(path_tail(lang), '_');
worse nvim/spell.c:        STRLCPY(lang, path_tail(spf_name), MAXWLEN + 1);
worse nvim/spellfile.c:  else if (vim_strchr(path_tail(wfname), '_') != NULL)
worse nvim/spellfile.c:        if (STRLEN(path_tail(innames[i])) < 5
worse nvim/tag.c:      filename = path_tail(buf);
worse nvim/tag.c:      && (p = path_tail(tag_fname)) != tag_fname) {
worse nvim/tag.c:      STRCPY(path_tail(buf), "tags");
```
